source/module_io/cube_io.h

```diff
@@ -26,42 +26,46 @@ void write_cube(
     const int precision = 11,
     const int out_fermi = 1); // mohan add 2007-10-17
 
+struct CubeInfo
+{
+    CubeInfo(
+        const std::vector<std::string>& comment,
+        const int natom,
+        const std::vector<double>& cel_pos,
+        const std::vector<int>& nvoxel,
+        const std::vector<std::vector<double>>& axis_vecs,
+        const std::vector<int>& atom_type,
+        const std::vector<double>& atom_charge,
+        const std::vector<std::vector<double>>& atom_pos,
+        const std::vector<double>& data,
+        const bool valid)
+        : comment(comment), natom(natom), cel_pos(cel_pos),
+        nvoxel(nvoxel), axis_vecs(axis_vecs),
+        atom_type(atom_type), atom_charge(atom_charge), atom_pos(atom_pos),
+        data(data), valid(valid)
+    {
+    };
 
-// when MPI:
-//      read file as order (ixy,iz) to data[ixy*nz+iz]
-// when serial:
-//      read file as order (ixy,iz) to data[iz*nxy+ixy]
-void read_cube_core_match(
-    std::ifstream &ifs,
-    const Parallel_Grid& pgrid,
-    const bool flag_read_rank,
-    double*const data,
-    const int nxy,
-    const int nz);
-
-void read_cube_core_mismatch(
-    std::ifstream &ifs,
-    const Parallel_Grid& pgrid,
-    const bool flag_read_rank,
-    double*const data,
-    const int nx,
-    const int ny,
-    const int nz,
-    const int nx_read,
-    const int ny_read,
-    const int nz_read);
+    const std::vector<std::string> comment = {};
+    const int natom = 0;
+    const std::vector<double> cel_pos = {};
+    const std::vector<int> nvoxel = {};
+    const std::vector<std::vector<double>> axis_vecs = {};
+    const std::vector<int> atom_type = {};
+    const std::vector<double> atom_charge = {};
+    const std::vector<std::vector<double>> atom_pos = {};
+    const std::vector<double> data = {};
+    const bool valid = false;
+};
 
-// when MPI:
-//      write data[ixy*nplane+iz] to file as order (ixy,iz)
-// when serial:
-//      write data[iz*nxy+ixy] to file as order (ixy,iz)
-void write_cube_core(
-    const Parallel_Grid& pgrid,
-    std::ofstream& ofs_cube,
-    const double*const data,
-    const int nxy,
-    const int nz,
-    const int n_data_newline);
+/// read the full data from a cube file 
+CubeInfo read_cube(const std::string& file);
```

Comment:
it is not the best practice. Please consider the following way which would be better:
First make the most general function:
```c++
void read_cube(const std::string& fn, 
               std::string& comment, 
               int& nx, 
               int& nz, 
               int& nz, 
               ModuleBase::Vector3<double>& dx, 
               ModuleBase::Vector3<double>& dy, 
               ModuleBase::Vector3<double>& dz,
               std::vector<int>& atomz,
               std::vector<double>& atomchg,
               std::vector<ModuleBase::Vector3<double>>& pos,
               std::vector<double>& data);
```
Then make an overload of it if you like the `CubeInfo` very much:
```c++
void read_cube(const std::string& fn, CubeInfo& info)
{
    std::string comment;
    int nx, ny, nz;
    // ...
    ModuleIO::read_cube(fn, comment, nx, ny, nz, ...);
    // then construct the CubeInfo instance
}
```
The first is the most general solution, so all developers will benefit from that. The second is based on your consideration, although you bind lifespans of all variables together, which I cannot see its benefit though. Do the same for `write_cube`:
```c++
void write_cube(const std::string& fn, 
                const std::string& comment, 
                const int nx, 
                const int nz, 
                const int nz, 
                const ModuleBase::Vector3<double>& dx, 
                const ModuleBase::Vector3<double>& dy, 
                const ModuleBase::Vector3<double>& dz,
                const std::vector<int>& atomz,
                const std::vector<double>& atomchg,
                const std::vector<ModuleBase::Vector3<double>>& pos,
                const std::vector<double>& data);
```

Reply:
or if you have special reason for not using `std::vector`, you make your kernel implementation like:
```c++
void read_cube(const std::string& fn, 
               std::string& comment, 
               int& nx, 
               int& nz, 
               int& nz, 
               ModuleBase::Vector3<double>& dx, 
               ModuleBase::Vector3<double>& dy, 
               ModuleBase::Vector3<double>& dz,
               int* atomz,
               double* atomchg,
               ModuleBase::Vector3<double>* pos,
               double* data);
```
in this way, the interface shown above will also be a wrapper of this, so implement the `std::vector` version like:
```c++
void read_cube(const std::string& fn, 
               std::string& comment, 
               int& nx, 
               int& nz, 
               int& nz, 
               ModuleBase::Vector3<double>& dx, 
               ModuleBase::Vector3<double>& dy, 
               ModuleBase::Vector3<double>& dz,
               std::vector<int>& atomz,
               std::vector<double>& atomchg,
               std::vector<ModuleBase::Vector3<double>>& pos,
               std::vector<double>& data)
{
    ModuleIO::read_cube(comment, nx, ny, nz, dx, dy, dz, atomz.data(), atomchg.data(), pos.data(), data.data());
}
```
.
Thanks.

Reply:
I've removed `CubeInfo` in the latest commit.